### PR TITLE
fix race condition at mkdir (#16258)

### PR DESCRIPTION
**Pull request by ewgRa**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #16258 |
| License | MIT |
| Doc PR |  |

### CODE REVIEW

**Comment by ewgRa** [comment]
I'm not sure is IOException in this case broke BC.

File: src/Symfony/Component/HttpKernel/Profiler/FileProfilerStorage.php [hunk 1]
```diff
@@ -42,7 +44,9 @@ public function __construct($dsn)
         $this->folder = substr($dsn, 5);
 
         if (!is_dir($this->folder)) {
-            mkdir($this->folder, 0777, true);
+            if (false === @mkdir($this->folder, 0777, true) && !is_dir($this->folder)) {
+                throw new IOException(sprintf('Unable to create the storage directory (%s).', $this->folder));
```
Comment (Tobion) [inline src/Symfony/Component/HttpKernel/Profiler/FileProfilerStorage.php]: The HttpKernel should not need to know the Filesystem Component exceptions. Simply use `\RuntimeException`

**Comment by Tobion** [comment]
The same problem is in line 145: https://github.com/ewgRa/symfony/blob/issue-16258-race-condition/src/Symfony/Component/HttpKernel/Profiler/FileProfilerStorage.php#L145

**Comment by Tobion** [comment]
Status: Needs Work

**Comment by ewgRa** [comment]
@Tobion Thanks for review

> The HttpKernel should not need to know the Filesystem Component exceptions
Where I can read more? It is quite strange, HttpKernel work with disk, but throw RuntimeException instead of IOException. And what If I want to find all code that work with IO? With IOException it is more obvious.

Status: Needs Review

**Comment by jvasseur** [comment]
@ewgRa The HttpKernel doesn't have any dependency on the filesystem component, using an exception would require to add it.

**Comment by ewgRa** [comment]
@jvasseur Thanks! Check composer at components and understand now that they can be used separately. Good to know.

**Comment by epilgrim** [comment]
I've seen this also happening at https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/HttpKernel/HttpCache/Store.php#L40

**Comment by ewgRa** [comment]
@epilgrim I think you are right. Checks added in HttpCache in three places where mkdir called.

**Comment by ewgRa** [comment]
Done. Previous build seems have random hhvm test fail.

**Comment by SpacePossum** [comment]
(for ref.: https://github.com/symfony/symfony/pull/14728 (please ignore))

**Comment by fabpot** [comment]
:+1: should be merged in 2.3 as this is a bug fix.

**Comment by fabpot** [comment]
Thank you @ewgRa.

File: src/Symfony/Component/HttpKernel/HttpCache/Store.php [hunk 2]
```diff
@@ -32,12 +32,16 @@ class Store implements StoreInterface
...
+     *
+     * @throws \RuntimeException
      */
     public function __construct($root)
     {
         $this->root = $root;
         if (!is_dir($this->root)) {
-            mkdir($this->root, 0777, true);
+            if (false === @mkdir($this->root, 0777, true) && !is_dir($this->root)) {
+                throw new \RuntimeException(sprintf('Unable to create the store directory (%s).', $this->root));
```
Comment (Tobion) [inline src/Symfony/Component/HttpKernel/HttpCache/Store.php]: typo: storage

**Pull request closed without merging**